snappy-core/src/main/scala/org/apache/spark/sql/store/CodeGeneration.scala

```diff
@@ -0,0 +1,261 @@
+        s"stmt.setBytes(${col + 1}, row.getBinary($col));"
+      case TimestampType =>
+        s"""
+          stmt.setTimestamp(${col + 1},
+            DateTimeUtils.toJavaTimestamp(row.getLong($col)));
+        """
+      case DateType =>
+        s"""
+          stmt.setDate(${col + 1},
+            DateTimeUtils.toJavaDate(row.getInt($col)));
+        """
+      case d: DecimalType =>
+        s"""
+          stmt.setBigDecimal(${col + 1},
+            row.getDecimal($col, ${d.precision}, ${d.scale}).toJavaBigDecimal);
+        """
+      case a: ArrayType =>
+        s"""
+          ArrayData arr = row.getArray($col);
+          if ($buffHolder == null) {
+            $buffHolder = new BufferHolder();
+          } else {
+            $buffHolder.reset();
+          }
+          ${GenerateUnsafeProjection.writeArrayToBuffer(ctx, "arr",
+            a.elementType, buffHolder)}
+          stmt.setBytes(${col + 1}, java.util.Arrays.copyOf($buffHolder.buffer,
+              $buffHolder.totalSize()));
+        """
+      case m: MapType =>
+        s"""
+          MapData map = row.getMap($col);
+          if ($buffHolder == null) {
+            $buffHolder = new BufferHolder();
+          } else {
+            $buffHolder.reset();
+          }
+          ${GenerateUnsafeProjection.writeMapToBuffer(ctx, "map",
+            m.keyType, m.valueType, buffHolder)}
+          stmt.setBytes(${col + 1}, java.util.Arrays.copyOf($buffHolder.buffer,
+              $buffHolder.totalSize()));
+        """
+      case s: StructType =>
+        s"""
+          InternalRow struct = row.getStruct($col, ${s.length});
+          if ($buffHolder == null) {
+            $buffHolder = new BufferHolder();
+          } else {
+            $buffHolder.reset();
+          }
+          ${GenerateUnsafeProjection.writeStructToBuffer(ctx, "struct",
+            s.fields.map(_.dataType), buffHolder)}
+          stmt.setBytes(${col + 1}, java.util.Arrays.copyOf($buffHolder.buffer,
+              $buffHolder.totalSize()));
+        """
+      case _ =>
+        s"stmt.setObject(${col + 1}, row.get($col, schema[$col].dataType()));"
+    }
+    (nonNullCode, s"stmt.setNull(${col + 1}, " +
+        s"${ExternalStoreUtils.getJDBCType(dialect, dataType)});")
+  }
+
+  private def compilePreparedUpdate(schema: Array[StructField],
+      dialect: JdbcDialect): CodeGeneration = {
+    val ctx = new CodeGenContext
+    val bufferHolder = ctx.freshName("bufferHolder")
+    val sb = new StringBuilder()
+    schema.indices.foreach { col =>
+      val (nonNullCode, nullCode) = getColumnSetterFragment(col,
+        schema(col).dataType, dialect, ctx, bufferHolder)
+      sb.append(
+        s"""
+          if (!row.isNullAt($col)) {
+            $nonNullCode
+          } else {
+            $nullCode
+          }
+        """)
+    }
+    val evaluator = new CompilerFactory().newScriptEvaluator()
+    evaluator.setClassName("io.snappydata.execute.GeneratedEvaluation")
```

Comment:
Yes, janino uses different classloader for each generated class. I guess this also helps with their GC with <=JDK7 that used PermGen (which used to be a problem with Derby generated classes).